Update tutorials and improve tutorial testing framework

jngrad opened the pull request:
Description of changes:
- tutorial 05 raspberry electrophoresis (fixes #3058)
   - plot the final trajectory in a matplotlib 3D plot
   - add test to check the raspberry traveled a longer distance over the axis of the E-field
   - test the sample .dat file as well
- tutorial 01 Lennard-Jones (fixes #3185)
   - plot and test the experimental RDF against the theoretical RDF
- tutorial 04 LB Part 4 solution script
   - added the solution script in the HTML output and fixed its test
- use AST to process Jupyter notebooks containing matplotlib code (fixes #3215)
- fix CMake dependency resolution for tutorials

jngrad committed:
Test tutorial 5

jngrad committed:
Change type of `ParticleHandle.fix` to bool

jngrad committed:
Update tutorials

Modernize code, fix broken HTML tags in figures, clean up
bibliographies, fix CMake logic.

jngrad committed:
Compare RDF with known curve (Morsali et al. 2005)

jngrad committed:
Plot raspberry trajectory

jngrad committed:
Increase test reliability

jngrad committed:
Improve processing of Jupyter notebook cells

Rewrite most of the regular expressions as AST node visitors to
improve the reliability of the importlib_wrapper. Better support
parsing of Jupyter cells containing matplotlib import statements,
which can lead to plots not being displayed in the HTML output.

jngrad committed:
Add test for html_runner

jngrad committed:
Replace regexes by AST (matplotlib, argparse)

jngrad committed:
Replace regexes by AST (random seeds, variables)

jngrad committed:
Fix CMake dependency graph for tutorials

jngrad committed:
Add solution to HTML output, simplify test logic

jngrad committed:
Fix errors in the solution and test

jngrad commented:
Further considerations:
* the theoretical RDF discussion in tutorial 01 may break the flow of the lecture and could be moved to the bottom of the notebook if necessary
* the new `html_runner` test fixture takes 7 seconds to run, which is noticeable when running a specific tutorial test with `make -j$(nproc) check_tutorials ARGS='-R tutorial_01'` for troubleshooting purposes
* new AST processing logic
   * enables a much more reliable handling of IPython magic commands that prevent the matplotlib command `plt.show()` from generating a picture in some tutorials (e.g. [04 Part 2](http://espressomd.org/html/tutorials_html/04-lattice_boltzmann/04-lattice_boltzmann_part2.html))
   * requires a lot of calls to `hasattr` (e.g. [importlib_wrapper.py#L388-L395](https://github.com/jngrad/espresso/blob/5c299fb8614fa11bbd1eafbe7a2e13ec1aafde07/testsuite/scripts/importlib_wrapper.py#L388-L395)), which could make `importlib_wrapper.py` fragile; please let me know if you're aware of a better strategy
   * doesn't use `ast.NodeTransformer` as we need to write the transformed code to disk to get clean tracebacks and create intermediate files for CMake dependency resolution

jngrad committed:
Refactor AST handling of espresso visualizers

Fixes a regression introduced by 1cc931f87759e4

KaiSzuttor commented on testsuite/scripts/tutorials/test_05-raspberry_electrophoresis.py:
```diff
@@ -17,15 +17,26 @@
 tutorial, skipIfMissingFeatures = importlib_wrapper.configure_and_import(
     "@TUTORIALS_DIR@/05-raspberry_electrophoresis/05-raspberry_electrophoresis.py",
-    gpu=True, num_iterations=10, num_steps_per_iteration=10)
+    gpu=True, box_l=20., E=0.2, num_iterations=150, num_steps_per_iteration=400)
```
the comment in the tutorial says that

> an electric field of 0.1 is the upper limit of the linear response regime for this model

you're setting it to 0.2 here

jngrad replied:
With an E-field < 0.2, the raspberry frequently travels a shorter distance along the axis field. Maybe the thermostat is too strong. Even with an E-field of 0.2, the random movement on the 2 other axes can still be larger than the E-field axis, requiring an annoying `min` function:
https://github.com/espressomd/espresso/blob/211fe9962130b8dadd3fb78e5aa5b983933d448e/testsuite/scripts/tutorials/test_05-raspberry_electrophoresis.py#L43
instead of the more correct solution
https://github.com/espressomd/espresso/blob/211fe9962130b8dadd3fb78e5aa5b983933d448e/testsuite/scripts/tutorials/test_05-raspberry_electrophoresis.py#L35-L36
I ran this test in a loop with different `E`, `num_iterations`, and `num_steps_per_iteration` to gather statistics and this inelegant solution is the one that didn't fail at random.

KaiSzuttor commented on testsuite/scripts/importlib_wrapper.py:
```diff
@@ -207,24 +210,86 @@ def set_random_seeds(code):
+    return dict(statements)
+
+
+def protect_ipython_magics(code):
```
please add docstrings for the free functions

jngrad replied:
done

KaiSzuttor commented on testsuite/scripts/tutorials/test_html_runner.py:
```diff
@@ -0,0 +1,123 @@
+        f_output = '@CMAKE_CURRENT_BINARY_DIR@/test_html_runner_notebook.run.ipynb'
+        f_script = '@CMAKE_CURRENT_BINARY_DIR@/test_html_runner_script.py'
+        # setup
+        if os.path.isfile(f_output):
```
is it mandatory for your test to rely on the hdd?

jngrad replied:
this requires `html_runner.py` being able to handle both filepaths or `io.StringIO` objects with a filepath attribute, in which case I might as well refactor `html_runner.py` as a class

KaiSzuttor commented on testsuite/scripts/tutorials/test_html_runner.py:
```diff
@@ -0,0 +1,123 @@
+import nbformat
+
+
+class HtmlRunner(ut.TestCase):
```
please add a short description of what you are testing how.

jngrad replied:
done

jngrad committed:
Move comparison with empirical RDF to the end

jngrad committed:
Move argparse to the top

Allows users to run the script with option `--help` and get the
documentation even if the script dependencies are not installed
or if CMake variables were not substituted.

jngrad committed:
Document wrappers and their tests

kodiakhq[bot] committed:
Merge branch 'python' into fix-3058
